File: backend/app/schemas/review.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.review import ReviewResult


ReviewStrategy = Literal["random", "spaced_repetition", "due_first"]


class ReviewItemOut(BaseModel):
    mistake_id: int
    title: str
    stem_markdown: str
    language: str
    difficulty: int
    category_name: str
    tag_names: list[str]
    shown_at: datetime


class ReviewRevealOut(BaseModel):
    mistake_id: int
    title: str
    stem_markdown: str
    wrong_answer_markdown: str
    correct_answer_markdown: str
    error_reason_markdown: str
    language: str
    difficulty: int
    category_name: str
    tag_names: list[str]


class ReviewProgressOut(BaseModel):
    completed: int
    total: int


class ReviewSessionStartIn(BaseModel):
    strategy: ReviewStrategy = "random"
    limit: int = Field(default=10, ge=1, le=50)


class ReviewSessionOut(BaseModel):
    id: int
    strategy: ReviewStrategy
    started_at: datetime
    total_count: int
    completed_count: int
    next_item: Optional[ReviewItemOut] = None


class ReviewNextOut(BaseModel):
    next_item: Optional[ReviewItemOut] = None
    progress: ReviewProgressOut


class ReviewSubmitIn(BaseModel):
    mistake_id: int
    user_result: ReviewResult
    shown_at: Optional[datetime] = None
    time_spent_ms: Optional[int] = Field(default=None, ge=0)
    note: Optional[str] = None


class ReviewLogOut(BaseModel):
    id: int
    mistake_id: int
    session_id: Optional[int]
    review_mode: str
    user_result: ReviewResult
    shown_at: datetime
    answered_at: Optional[datetime]
    time_spent_ms: Optional[int]
    note: str

    model_config = ConfigDict(from_attributes=True)


class ReviewSubmitOut(ReviewLogOut):
    progress: ReviewProgressOut


class ReviewResultCountsOut(BaseModel):
    again: int = 0
    hard: int = 0
    good: int = 0
    easy: int = 0


class ReviewSummaryOut(BaseModel):
    total_count: int
    completed_count: int
    result_counts: ReviewResultCountsOut
    duration_ms: int


class ReviewCapabilityOut(BaseModel):
    ai_analysis_enabled: bool
    model: Optional[str] = None


class ReviewDueCountOut(BaseModel):
    due_count: int
    as_of: datetime

File: frontend/src/types/review.ts
export type ReviewStrategy = "random" | "spaced_repetition" | "due_first";
export type ReviewResult = "again" | "hard" | "good" | "easy";

export interface ReviewItem {
  mistake_id: number;
  title: string;
  stem_markdown: string;
  language: string;
  difficulty: number;
  category_name: string;
  tag_names: string[];
  shown_at: string;
}

export interface ReviewReveal {
  mistake_id: number;
  title: string;
  stem_markdown: string;
  wrong_answer_markdown: string;
  correct_answer_markdown: string;
  error_reason_markdown: string;
  language: string;
  difficulty: number;
  category_name: string;
  tag_names: string[];
}

export interface ReviewProgress {
  completed: number;
  total: number;
}

export interface ReviewSessionStartPayload {
  strategy: ReviewStrategy;
  limit?: number;
}

export interface ReviewSession {
  id: number;
  strategy: ReviewStrategy;
  started_at: string;
  total_count: number;
  completed_count: number;
  next_item: ReviewItem | null;
}

export interface ReviewNextResponse {
  next_item: ReviewItem | null;
  progress: ReviewProgress;
}

export interface ReviewSubmitPayload {
  mistake_id: number;
  user_result: ReviewResult;
  shown_at?: string;
  time_spent_ms?: number;
  note?: string;
}

export interface ReviewLog {
  id: number;
  mistake_id: number;
  session_id: number | null;
  review_mode: string;
  user_result: ReviewResult;
  shown_at: string;
  answered_at: string | null;
  time_spent_ms: number | null;
  note: string;
}

export interface ReviewSubmitResponse extends ReviewLog {
  progress: ReviewProgress;
}

export interface ReviewResultCounts {
  again: number;
  hard: number;
  good: number;
  easy: number;
}

export interface ReviewSummary {
  total_count: number;
  completed_count: number;
  result_counts: ReviewResultCounts;
  duration_ms: number;
}

export interface ReviewCapability {
  ai_analysis_enabled: boolean;
  model?: string;
}

export interface ReviewDueCount {
  due_count: number;
  as_of: string;
}

export type AiAnalysisStreamStatus = "idle" | "ready" | "streaming" | "completed" | "error";
